v2/api/devices/v1api20210702/structure.txt

```diff
@@ -0,0 +1,841 @@
+// Code generated by azure-service-operator-codegen. DO NOT EDIT.
+github.com/Azure/azure-service-operator/v2/api/devices/v1api20210702
+├── APIVersion: Enum (1 value)
+│   └── "2021-07-02"
+├── IotHub: Resource
+│   ├── Owner: github.com/Azure/azure-service-operator/v2/api/resources/v1apiv20191001.ResourceGroup
+│   ├── Spec: Object (9 properties)
+│   │   ├── AzureName: string
+│   │   ├── Etag: *string
+│   │   ├── Identity: *Object (2 properties)
+│   │   │   ├── Type: *Enum (4 values)
+│   │   │   │   ├── "None"
+│   │   │   │   ├── "SystemAssigned"
+│   │   │   │   ├── "SystemAssigned, UserAssigned"
+│   │   │   │   └── "UserAssigned"
+│   │   │   └── UserAssignedIdentities: Object (1 property)[]
+│   │   │       └── Reference: genruntime.ResourceReference
+│   │   ├── Location: *string
+│   │   ├── OperatorSpec: *Object (1 property)
+│   │   │   └── Secrets: *Object (10 properties)
+│   │   │       ├── DevicePrimaryKey: *genruntime.SecretDestination
+│   │   │       ├── DeviceSecondaryKey: *genruntime.SecretDestination
+│   │   │       ├── IotHubOwnerPrimaryKey: *genruntime.SecretDestination
+│   │   │       ├── IotHubOwnerSecondaryKey: *genruntime.SecretDestination
+│   │   │       ├── RegistryReadPrimaryKey: *genruntime.SecretDestination
+│   │   │       ├── RegistryReadSecondaryKey: *genruntime.SecretDestination
+│   │   │       ├── RegistryReadWritePrimaryKey: *genruntime.SecretDestination
+│   │   │       ├── RegistryReadWriteSecondaryKey: *genruntime.SecretDestination
+│   │   │       ├── ServicePrimaryKey: *genruntime.SecretDestination
+│   │   │       └── ServiceSecondaryKey: *genruntime.SecretDestination
+│   │   ├── Owner: *genruntime.KnownResourceReference
+│   │   ├── Properties: *Object (19 properties)
+│   │   │   ├── AllowedFqdnList: string[]
+│   │   │   ├── AuthorizationPolicies: Object (4 properties)[]
+│   │   │   │   ├── KeyName: *string
+│   │   │   │   ├── PrimaryKey: *string
```

Comment:
Is this used to configure the primary key of the policy or something? Or is it readonly?

Feels like either this whole `AuthorizationPolicies` thing should be marked readonly (if it is?), or else the Primary/Secondary key fields should be marked with `x-ms-secret`

Reply:
Thanks for pointing out. Only PrimaryKey and SecondaryKey fields should be marked as read-only and are generated by azure. That said, user can specify their own custom auth policy with the keyName and accessRights and Azure would generate the keys for them. However, the only problem here is, how would we expose those keys to the users once its created by the user?

Reply:
Isn't this what AzureGeneratedSecrets is about? Though, we specify that for the Status of the resource, and it looks like there's a whole array here which is ... fun!